File: vivaldi/lib/simulation/vector.ex
defmodule Vivaldi.Simulation.Vector do
  @moduledoc """
  Vector functions. All functions assume dimensions match.
  """

  def distance(p1, p2) do
    diff(p1, p2)
    |> magnitude()
  end

  def diff(p1, p2) do
    Enum.zip(p1, p2)
    |> Enum.map(fn {a, b} -> (a - b) end)
  end

  def add(p1, p2) do
    Enum.zip(p1, p2)
    |> Enum.map(fn {a, b} -> (a + b) end)
  end

  def magnitude(p1) do
    p1
    |> Enum.map(&(&1 * &1))
    |> Enum.reduce(&(&1 + &2))
    |> :math.sqrt()
  end

  def scale(p1, factor) do
    Enum.map(p1, fn a -> a * factor end)
  end

  def zero(dimension) do
    Enum.map(1..dimension, fn _ -> 0 end)
  end

  def rand(dimension) do
    Enum.map(1..dimension, fn _ -> :rand.uniform - 0.5 end)
  end

  @doc """
  According to the paper:
  > Because all nodes start at the same location, Vivaldi must separate them somehow.
  > Vivaldi does this by defining u(0) to be a unit-length vector in a randomly chosen direction
  When p1 and p2 are non-zero vectors, then the conventional definition of a unit vector is used.
  """
  def unit_vector_at(p1, p2) do
    dimension = Enum.count(p1)
    if p1 == zero(dimension) and p2 == zero(dimension) do
      u = rand(dimension)
      scale(u, 1 / magnitude(u))
    else
      if p1 == p2 do
        Enum.map(1..dimension, fn _ -> 0 end)
      else
        d = diff(p1, p2)
        mag = magnitude(d)
        scale(d, 1 / mag)
      end
    end
  end
end

File: visualization/src/vector.js
export default {
  add(p1, p2) {
    return zip(p1, p2).map(value => value[0] + value[1])
  },

  diff(p1, p2) {
    return zip(p1, p2).map(value => value[0] - value[1])
  },

  distance(p1, p2) {
    return this.magnitude(this.diff(p1, p2));
  },

  magnitude(p1) {
    return Math.sqrt(p1.map(value => value * value).reduce((a, b) => a + b))
  },

  rand(dimension) {
    let a = new Array(dimension)
    for (let i = 0; i < dimension; i++) {
      a[i] = Math.random() - 0.5
    }
    return a
  },

  scale(p1, factor) {
    return p1.map(value => value * factor)
  },

  unitVectorAt(p1, p2) {
    if (this.isZero(p1) && this.isZero(p2)) {
      const r = this.rand(p1.length)
      return this.scale(r, 1 / this.magnitude(r))
    } else if (areEqual(p1, p2)) {
      return this.zero(p1.length)
    } else {
      const d = this.diff(p1, p2)
      const mag = this.magnitude(d)
      return this.scale(d, 1 / mag)
    }
  },

  zero(dimension) {
    let a = new Array(dimension);
    for (let i = 0; i < dimension; i++) {
      a[i] = 0
    }
    return a
  },

  isZero(p1) {
    for (let value of p1) {
      if (value !== 0) {
        return false;
      }
    }
    return true;
  }
}

function areEqual(p1, p2) {
  for (let [a, b] of zip(p1, p2)) {
    if (a !== b) {
      return false
    }
  }
  return true
}

function zip(p1, p2) {
  return p1.map((value, index) => [value, p2[index]])
}
